### PR TITLE
refactor: port tensorflow model to pytorch

### LINKED ISSUE #8: Implemetation of segmentation model in pytorch
Pytorch implementation of the current model:
- [ ] UNet model
- [ ] Weights and biases logging

### PR DESCRIPTION
**Pull request by aledelunap**
closes #8 

The pytorch port includes the following:
- dataset.py -> Dataset implementation with pytorch lightning 
- model.py -> UNet implementation with pytorch lightning. The LWBNA UNet implementation is pending, the current pull request includes a submodule for the UNet architecture
- training.py -> Training stage with pytorch lightning, model is saved as *.pth

### CODE REVIEW

File: src/dataset.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    def __getitem__(self, idx):
+        if torch.is_tensor(idx):
+            idx = idx.tolist()
+
+        fiber_file = self.fiber_dic[idx]
+
+        x, y = get_example(self.data_dir, fiber_file)
+
+        x = torch.tensor(x)
+        y = torch.tensor(y)[None]
```
Comment (fcossio) [inline src/dataset.py]: I am ignorant of why the None is used. What does it mean?

File: src/dataset.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+from PIL import Image, ImageOps
+
+
+def get_fibers_dictionary(data_dir: str, stage: str):
+
+    fibers_dict = {}
+    idx = 0
+
+    for file in os.listdir(data_dir):
+        if file[: len(stage)] == stage and file[-6:] == "params":
```
Comment (fcossio) [inline src/dataset.py]: Consider using .startswith(stage) and .endswith("params") that are easier to read.

File: src/dataset.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+import torch
+from torch.utils.data import DataLoader
+import pytorch_lightning as pl
+from torchvision import transforms
+
+from PIL import Image, ImageOps
+
+
+def get_fibers_dictionary(data_dir: str, stage: str):
```
Comment (fcossio) [inline src/dataset.py]: Do I understand correctly that it returns a dictionary mapping the images of a data split? Why is it needed to use a dictionary if the keys are integers?

File: src/dataset.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    point, selected_seg = select_point_and_fiber(seg)
+    point_channel = np.zeros_like(im, dtype=np.float32)
+    point_channel[0, point[0], point[1]] = 1.0
+    x = np.concatenate([im, point_channel])
+    y = selected_seg
+    return x, y
+
+
+class FiberDataset(torch.utils.data.Dataset):
+    def __init__(self, stage: str, fiber_dic: dict, data_dir: str):
```
Comment (fcossio) [inline src/dataset.py]: How is fiber_dict supposed to look like? Maybe add it in the docstring

File: src/dataset.py (hunk 1 above)
Comment (fcossio) [inline src/dataset.py]: Why is `[None]` needed?

File: src/dataset.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        else:
+            test_dict = get_fibers_dictionary(self.data_dir, "test")
+
+            self.fiber_test = FiberDataset(
+                "test",
+                test_dict,
+                self.data_dir,
+            )
+
+    def train_dataloader(self: pl.LightningDataModule):
```
Comment (fcossio) [inline src/dataset.py]: I don't think self needs to be typed, as it can be deduced from the class, but maybe I'm wrong?

File: src/training.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
+import sys
+from argparse import ArgumentParser
+import inspect
+
+import torch
+import wandb
+from pytorch_lightning import Trainer, seed_everything
+
+from model import UNet
+from dataset import FiberDataModule
```
Comment (fcossio) [inline src/training.py]: Maybe this can go inside the fiberdiameter folder so we can import it as fiberdiameter.dataset and it becomes easier to test in the future

File: src/training.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        "--data_dir", type=str, default="artifacts/rendered-fibers-medium:v0"
+    )
+    parser.add_argument("--num_workers", type=int, default=12)
+    parser.add_argument("--limit_train_batches", type=float, default=0.1)
+    parser.add_argument("--shuffle", type=bool, default=True)
+    args = parser.parse_args(raw_args)
+    return args
+
+
+def init_from_valid_args(cls, args):
```
Comment (fcossio) [inline src/training.py]: Curiosity question: what is the use of this module?

File: .gitmodules [hunk 8]
```diff
@@ -0,0 +1,3 @@
+[submodule "src/unet_pytorch"]
+	path = src/unet_pytorch
+	url = git@github.com:jaxony/unet-pytorch.git
```
Comment (fcossio) [inline .gitmodules:3]: Not v important: Is it possible to specify the hash of the commit hash that is being used?

File: pyproject.toml [hunk 9]
```diff
@@ -2,7 +2,12 @@
 name = "fiberdiameter"
 version = "0.1.1"
 description = "Measure the diameter of fibers 10x faster!"
-authors = ["Fernando Cossio", "Yael Suarez", "Subhadeep Chowdhury", "Luis Cossio", "Ale de Luna"]
+authors = [
```
Comment (fcossio) [inline pyproject.toml:5]: please add the required dependencies to the package with

`poetry add <X>` # if it will be used by the package
`poetry add --group dev <X>` # if it will be used for dev only

**Review by fcossio — CHANGES_REQUESTED** [review]
Great work! I have left some comments in the files.

Please don't delete the files that are being deleted and put the model and dataloaders inside the dataset and segmentation folders.

File: src/dataset.py (hunk 3 above)
Comment (aledelunap) [reply src/dataset.py:14]: the integers are needed in order to use the dataloader directly, so the dictionary keeps track of the mapping between the idx (integer) and the respective image

File: src/dataset.py (hunk 2 above)
Comment (aledelunap) [reply src/dataset.py]: perfect!

File: src/dataset.py (hunk 1 above)
Comment (aledelunap) [reply src/dataset.py:84]: It just adds a dimension to end up with the shape (1, n, n) to prepare for batching

File: src/dataset.py (hunk 3 above)
Comment (aledelunap) [reply src/dataset.py]: I will delete this function, I now see there is no need for a dictionary

File: src/dataset.py (hunk 4 above)
Comment (aledelunap) [reply src/dataset.py:65]: fiber_dict will be removed

File: src/dataset.py (hunk 1 above)
Comment (aledelunap) [reply src/dataset.py]: to add an extra dimension for batching (1, n, n), I will refactor it to just move the axis of the original image instead of squeezing it

Comment (aledelunap) [reply src/dataset.py]: This is already answered in another comment.

File: src/training.py (hunk 7 above)
Comment (aledelunap) [reply src/training.py]: it initializes the class cls with the respective arguments contained in args since the parser receives as input variables for different classes

File: src/dataset.py (hunk 5 above)
Comment (aledelunap) [reply src/dataset.py]: the course types it as such (I don't know if it a standard thing), I can remove it

File: pyproject.toml (hunk 9 above)
Comment (aledelunap) [reply pyproject.toml:5]: done

**Review by fcossio — APPROVED** [review]
Nice! great first PR
